### ISSUE TITLE
Improve parallel execution

### ISSUE THREAD

**Issue by aztech-dev** [padraic/humbug#124]
Suggestion is to add a CLI option to specify number of parallel runs.

Additionally, changing the number of parallel runs has side effects on test outcome (ie. some errors become escapes from what I briefly noticed).

And lastly, parallel execution is not optimal as execution is split into groups during the execution of tests for a given mutable -- worst (theoretical) case being 1 mutation per mutable file would result in non-parallel execution.

**Comment by rollenes** [comment]
Running test suites in parallel is big problem as your tests may rely on some storage. 
For example: Test creates files on disk.
Running the same test in parallel may cause that one test creates file, and the same test run in parallel may not create it as it already exists. 
In that case second test fails or reports error.
* If it reports error you have to check what was reason of error - you may check what was a problem in that case, but it takes time to understand what happened. 
* If it fails Humbug finds this mutation as killed - and you don`t want that behavior.

I don`t see any simple solution for this kind of issue except running it on separate machines. But it also has weak points as running for example DB integration suites. Those suites may rely on the same database configuration...

**Comment by rollenes** [comment]
What if humbug may run not mutations in parallel, but the same tests for different mutations in paralle?
For example: 
```
1.mutant
  1.test
  2.test
2 mutant
  1.test
  2.test

And if it were combined in:
1.process runs:
  1.mutant - 1.test
  2.mutant - 1.test
2.process runs:
  1.mutant - 2.test
  2.mutant - 2.test
```
So grouping tests may be solution.

**Comment by padraic** [comment]
Parallel mutants are actually worse since it increases the odds of race conditions on shared test dependencies (e.g. writes/reads to same /tmp files). Parallel with a single mutation is more possible. It can be done, but would need another adapter for the parallel runner. If we go the route of running a localised phpunit (and not the bundled copy) that would simplify this route considerably.

It would be an entirely optional feature - many test suites are simply too interdependent to safely run in parallel or don't cleanup resources correctly. If you check the phpunit-extensions package, there's a random order listener in there intended to ease the transition by running tests in random orders to locate these problems.

**Comment by aztech-dev** [comment]
Yup, I forgot to mention the testsuite I tried it on normally has no dependencies on databases or filesystem -- though I should probably double-check.

I originally considered the same solution as @rollenes, partitioning at the mutable file level, but I see what you mean, so probably not a good solution.

Looks like there is no magical solution to rule them all, but being able to pass an option to change number of parallel runs would already be a start, and should of course default to 1 when omitted.

I'll play around a bit with the code before pushing further for any changes in parallel mode.

**Comment by padraic** [comment]
For PHPUnit at least, there is https://github.com/brianium/paratest, which is what I meant by having a separate adapter (not necessarily needs to be separate if the option can be integrated with the existing phpunit one). It's already used so best to rely on it where possible.

The existing ParallelGroup stuff was originally heading down this road, but why reinvent the wheel? There are already folk giving paratest a lot of attention and maintenance so let's leverage off that.
